### ISSUE TITLE
Leafs with no training events in them

### ISSUE THREAD

**Issue by Denisevi4** [dmlc/xgboost#3826]
Very rarely I incur this instance of having leafs in the model that have cover=0.

Here is a dump of such a tree (sorry. don't know how to format it)

0:[C19<-0.00326049794] yes=1,no=2,missing=1,gain=6.68777793e-05,cover=205821
	1:[C49<13.9841671] yes=3,no=4,missing=3,gain=0.000157496921,cover=20660
		3:[C104<1.53036571] yes=7,no=8,missing=7,gain=0.000128577725,cover=18218
			7:[C10<39774.3789] yes=15,no=16,missing=15,gain=0.000146228616,cover=16352
				15:leaf=6.29458373e-05,cover=8496
				16:leaf=-0,cover=7856
			8:[C147<18.836647] yes=17,no=18,missing=17,gain=0.000132655929,cover=1866
				17:leaf=-0,cover=1138
				18:leaf=-0.000121151919,cover=728
		4:[C140<7.49999973e-09] yes=9,no=10,missing=9,gain=0.000101037891,cover=2442
			9:leaf=-0,cover=1434
			10:[C49<13.9841671] yes=19,no=20,missing=19,gain=6.39266946e-05,cover=1008
				**19:leaf=-nan,cover=0**
				20:leaf=-9.8133547e-05,cover=1008
	2:[C71<3.32500008e-07] yes=5,no=6,missing=5,gain=9.50862304e-05,cover=185161
		5:[C127<0.237853229] yes=11,no=12,missing=11,gain=0.000120236931,cover=180340
			11:[C102<1.97036135] yes=21,no=22,missing=21,gain=0.000182247895,cover=173194
				21:leaf=-0,cover=160209
				22:leaf=-3.38961036e-05,cover=12985
			12:[C146<6.92263126] yes=23,no=24,missing=23,gain=0.000333737611,cover=7146
				23:leaf=-3.61185957e-05,cover=1833
				24:leaf=7.79525362e-05,cover=5313
		6:[C16<0.0269251391] yes=13,no=14,missing=13,gain=0.000221637121,cover=4821
			13:[C115<0.623490334] yes=25,no=26,missing=25,gain=2.03016607e-05,cover=2146
				25:leaf=-0,cover=1280
				26:leaf=4.27486339e-05,cover=866
			14:[C13<0.00162636698] yes=27,no=28,missing=27,gain=0.000202318741,cover=2675
				27:leaf=-0.000146710634,cover=1902
				28:leaf=-0,cover=773

Node number 19 has cover=0, which means that there are no training events in that node. Current xgboost master puts leaf=-nan for the value in the node (which is how I detected the problem because my code was crashing trying to read it), whereas xgboost 0.71 produces the same tree, but puts leaf=0 in there but still has cover=0.

This happens if I use tree_method exact or approx. I have min_child_weight=20.

My question is how is this possible? 

Is this a bug or known issue?

If this is a bug I can provide a data sample and a simple way to reproduce it (run just a single tree).

I use CentOs7.

**Comment by hcho3** [comment]
Can you provide a data sample that I can run and reproduce the issue? This looks like a bug.

**Comment by Denisevi4** [comment]
Here is the dataset.

The code to reproduce it:

    import xgboost
    import pandas as pd

    df = pd.read_hdf("weird.h5")
    dtrain = xgboost.DMatrix(df.drop(columns=["TARGET"]), label=df["TARGET"])

    pars = {'base_score': 0.0,
     'booster': 'gbtree',
     'eta': 1.0,
     'eval_metric': 'rmse',
     'gamma': 0.0,
     'max_depth': 4,
     'min_child_weight': 20.0,
     'objective': 'reg:linear',
     'reg_alpha': 0.35,
     'reg_lambda': 0.0,
     'silent': 1,
     "tree_method": "exact"
    }
    # fit one tree
    bst_pr = xgboost.train(pars, dtrain, 1)

    # dump the model
    dd = bst_pr.get_dump(with_stats=True)
    print dd[0]




[data.zip](https://github.com/dmlc/xgboost/files/2511669/data.zip)

**Comment by Denisevi4** [comment]
The leaf 17 would be with cover=0.

If you look at the split nodes, nodes 1 and 10 have an identical split, which leads to that cover=0 node.

0:C2 < -0.00652099587
**1:C1 < 27.9683342**
2:C1 < 23.9683456
3:C2 < -0.0164765697
4:C0 < 1.49999995e-08
5:C1 < 19.1684074
6:C1 < 26.7018356
8:C0 < 2.30000001e-08
**10:C1 < 27.9683342**
11:C0 < 9.99999972e-10
12:C2 < 0.006448247
13:C2 < 0.0112791462
14:C0 < 0.524977624

**Comment by Denisevi4** [comment]
Also, if you make max_depth=5, there will be another node with cover=0.

**Comment by Denisevi4** [comment]
@hcho3 did you have a chance to look at the problem?

**Comment by hcho3** [comment]
@Denisevi4 Will investigate this problem soon.

**Comment by cowden** [comment]
@hcho3 @Denisevi4 
Hi,

I came across this same behavior as well, and I was able to track down the cause to a large concentration of data at a single floating point number (or at least the distribution is below the precision of floats).  The exact greedy method can't resolve the split between observed values, so it gets rounded, and in this situation the "<" logic in scoring a tree differs from the sorting and splitting done in the search.

https://github.com/dmlc/xgboost/pull/5010
I opened a pull request that implements a simple check.  There may be other ways to protect from redundant splits and 0 cover nodes in pruning or implementing "<=" logic, but I thought those likely require more substantial changes to the package.

Here are the spilts found using the proposed patch.
0:C2 < -0.00652099587
1: **C1 < 27.9683361**
3:C0 < 526.223633
7:C0 < 2.30000001e-08
4:C0 < 1.49999995e-08
10:C2 < -0.00678067794
2:C1 < 23.9683456
5:C1 < 19.1684093
11:C0 < 9.99999972e-10
12:C2 < 0.006448247
6:C1 < 26.7018356
13:C2 < 0.0112791462
14:C0 < 0.524977624
